libsolidity/interface/CompilerStack.cpp

```diff
@@ -194,7 +217,30 @@ bool CompilerStack::parse()
 				else
 					noErrors = false;
 
-				m_contracts[contract->name()].contract = contract;
+				if (m_contracts.find(contract->fullyQualifiedName()) != m_contracts.end())
+				{
+					const ContractDefinition* existingContract = m_contracts.find(contract->fullyQualifiedName())->second.contract;
+
+					if (contract != existingContract)
+					{
+						auto err = make_shared<Error>(Error::Type::DeclarationError);
```

Comment:
I don't get it, all contracts in your example have different fully qualified names, so we should not get to line 226.